plan.lua

```diff
@@ -213,100 +213,174 @@ function plan_class:get_random_plan_pos()
+	local function read_s16(fi)
+		return string.byte(fi:read(1)) * 256 + string.byte(fi:read(1))
+	end
 
-	-- Minetest Schematics
-	if mts_format then
-		-- thanks to sfan5 for this advanced code that reads the size from schematic files
-		local function read_s16(fi)
-			return string.byte(fi:read(1)) * 256 + string.byte(fi:read(1))
-		end
+	-- make sure those are the first 4 characters, otherwise this might be a corrupt file
+	if file:read(4) ~= "MTSM" then
+		file.close()
+		return
+	end
+	local result = {type="mts"}
+	-- advance 2 more characters
+	local version = read_s16(file) --f:read(2)
+	result["version"] = version
+		-- the next characters here are our size, read them
+	local size_x = read_s16(file)
+	local size_y = read_s16(file)
+	local size_z = read_s16(file)
+	result["size"] = {x=size_x, y=size_y, z=size_z}
+	-- read the slice probability for each y value that was introduced in version 3
+	if version >= 3 then
+		--TODO: Read probability
+		result["probability"] = file:read(size_y)
+	end
+
+	-- read node names
+	local node_name_count = read_s16(file)
+	local node_names = {}
+	for _ = 1, node_name_count do
+		local name_length = read_s16(file)
+		local name_text   = file:read(name_length)
+		table.insert(node_names, name_text)
+	end
+	result["node_name_count"] = node_name_count
+	result["node_names"] = node_names
+	result["data"] = file:read( "*all" );
+	file.close()
+	return result
+end
 
-		-- make sure those are the first 4 characters, otherwise this might be a corrupt file
-		if file:read(4) ~= "MTSM" then
-			return
-		end
-		-- advance 2 more characters
-		local version = read_s16(file) --f:read(2)
-			-- the next characters here are our size, read them
-		local size_x = read_s16(file)
-		local size_y = read_s16(file)
-		local size_z = read_s16(file)
-		-- read the slice probability for each y value that was introduced in version 3
-		if version >= 3 then
-			--TODO: Read probability
-			file:read(size_y)
-		end
+--------------------------------------
+-- return WE Data from Minetest WorldEdit Schematic file
+--------------------------------------
+function plan_class:get_from_we_file(filename)
+	local ext = get_ext_name(filename) -- .we or .wem
+	if ext == nil then filename = filename .. '.we' end
 
-		-- read node names
-		local node_name_count = read_s16(file)
-		local node_names = {}
-		local air_id = 0;
-		for i = 1, node_name_count do
-			local name_length = read_s16(file)
-			local name_text   = file:read(name_length)
-			if replacements and replacements[name_text] then
-				name_text = replacements[name_text]
-			end
-			table.insert(node_names, name_text)
-			if name_text == "air" then
-				air_id = i
-			end
-		end
+	local file = io.open( filename, 'r' )
+	if not file then
+		dprint("error: could not open file \"" .. filename .. "\"")
+		return
+	end
+	local result = {type = "we", data = schemlib.worldedit_file.load_schematic(file:read("*a"))}
+	file.close()
```

Comment:
See the `file.close()` call is wrong. Right call is `file.close(file)` or better syntax: `file:close()`.
Please consider this change